odbc/src/api.rs

```diff
@@ -1,17 +1,83 @@
+use crate::handles::{
+    Connection, ConnectionState, Env, EnvState, MongoHandle, Statement, StatementState,
+};
 use odbc_sys::{
     BulkOperation, CDataType, Char, CompletionType, ConnectionAttribute, Desc, DriverConnectOption,
     EnvironmentAttribute, FetchOrientation, HDbc, HDesc, HEnv, HStmt, HWnd, Handle, HandleType,
     InfoType, Integer, Len, Nullability, ParamType, Pointer, RetCode, SmallInt, SqlDataType,
     SqlReturn, StatementAttribute, ULen, USmallInt, WChar,
 };
+use std::sync::RwLock;
 
 #[no_mangle]
 pub extern "C" fn SQLAllocHandle(
-    _handle_type: HandleType,
-    _input_handle: Handle,
-    _output_handle: *mut Handle,
-) -> SqlReturn {
-    unimplemented!()
+    handle_type: HandleType,
+    input_handle: Handle,
+    output_handle: *mut Handle,
+) -> SqlReturn {
+    match sql_alloc_handle(handle_type, input_handle as *mut _, output_handle) {
+        Ok(_) => SqlReturn::SUCCESS,
+        Err(_) => SqlReturn::INVALID_HANDLE,
+    }
+}
+
+fn sql_alloc_handle(
+    handle_type: HandleType,
+    input_handle: *mut MongoHandle,
+    output_handle: *mut Handle,
+) -> Result<(), ()> {
+    match handle_type {
+        HandleType::Env => {
+            let env = RwLock::new(Env::with_state(EnvState::Allocated));
+            let mh = Box::new(MongoHandle::Env(env));
+            unsafe {
+                *output_handle = Box::into_raw(mh) as *mut _;
+            }
+            Ok(())
+        }
+        HandleType::Dbc => {
+            let conn = RwLock::new(Connection::with_state(
+                input_handle,
+                ConnectionState::Allocated,
+            ));
+            // input handle cannot be NULL
+            if input_handle == std::ptr::null_mut() {
+                return Err(());
+            }
+            // input handle must be an Env
+            let env = unsafe { (*input_handle).as_env()? };
```

Comment:
If the check of the input type was the first thing done in the block, instead of creating a new connection which will be discarded when the input handle check fails, would it save us some time by not having to allocate and free memory? This is more of a general question, so that I can have an idea of the impact of this kind of things. I generally try to fail as early as possible to avoid unnecessary work because in Java for example creating the new Connection instance would take (some minimal) time and have to be garbage collected soon after (which also take some time and if too many useless instance of Connection are created by a dumb programm, the GB could be running more cycles than expected), but I am not sure of the incidence in Rust.

Reply:
Oh, I agree. I'll check that out. This was kind of silly.